### PR TITLE
gh-117607: Speedup `os.path.relpath()`

### PR DESCRIPTION
**Pull request by nineteendo**
* Issue: gh-117607

### CODE REVIEW

File: Lib/ntpath.py [hunk 1]
```diff
@@ -785,33 +785,36 @@ def realpath(path, *, strict=False):
...
         start_drive, _, start_rest = splitroot(start_abs)
         path_drive, _, path_rest = splitroot(path_abs)
         if normcase(start_drive) != normcase(path_drive):
             raise ValueError("path is on mount %r, start on mount %r" % (
                 path_drive, start_drive))
 
-        start_list = [x for x in start_rest.split(sep) if x]
-        path_list = [x for x in path_rest.split(sep) if x]
+        start_list = start_rest.split(sep) if start_rest else []
+        path_list = path_rest.split(sep) if path_rest else []
```
Comment (ssweber) [inline Lib/ntpath.py:815]: just looking at this, these don't exactly do the same thing, is that OK. (first filters list).

Comment (nineteendo) [reply Lib/ntpath.py:815]: In theory the segments of `start_rest` can only be empty when `start_rest` is empty, as `start_abs` is already normalised. But some tests are failing, so I need to verify this.

Comment (nineteendo) [reply Lib/ntpath.py:815]: Appears to be working now.

**Comment by nineteendo** [comment]
@barneygale, could you review this?

**Comment by nineteendo** [comment]
@serhiy-storchaka, do you think this can break? Or is it fine?

File: Misc/NEWS.d/next/Core and Builtins/2024-04-07-18-42-09.gh-issue-117607.C978BD.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Speedup :func:`os.path.relpath()`.
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-07-18-42-09.gh-issue-117607.C978BD.rst]: Nit: `:func:` will add the parentheses for you.

```suggestion
Speedup :func:`os.path.relpath`.
```

**Comment by nineteendo** [comment]
@eryksun, could you run a benchmark for `os.path.relpath()` on your machine? Otherwise it might undersell the performance improvement.

**Comment by nineteendo** [comment]
Could anyone do a benchmark of Unix? `os.getcwd()` takes way too long on my machine.

**Comment by eryksun** [comment]
> @eryksun, could you run a benchmark for `os.path.relpath()` on your machine? Otherwise it might undersell the performance improvement.

Windows - ntpath - 1.407-1.565 speedup

```text
> .\python_main -m timeit -s "import ntpath" "ntpath.relpath('C:foo', start='C:/Users/account')"
50000 loops, best of 5: 6.7 usec per loop
> .\python_work -m timeit -s "import ntpath" "ntpath.relpath('C:foo', start='C:/Users/account')"
50000 loops, best of 5: 4.28 usec per loop

> .\python_main -m timeit -s "import ntpath" "ntpath.relpath('C:/', start='C:/Users/account')"
50000 loops, best of 5: 4.84 usec per loop
> .\python_work -m timeit -s "import ntpath" "ntpath.relpath('C:/', start='C:/Users/account')"
100000 loops, best of 5: 3.44 usec per loop
```

Linux - posixpath - 1.137-1.189 speedup

```text
$ ./python_main -m timeit -s "import posixpath" "posixpath.relpath('foo', start='/home/account')"
100000 loops, best of 5: 3.14 usec per loop
$ ./python_work -m timeit -s "import posixpath" "posixpath.relpath('foo', start='/home/account')"
100000 loops, best of 5: 2.64 usec per loop

$ ./python_main -m timeit -s "import posixpath" "posixpath.relpath('/', start='/home/account')"
200000 loops, best of 5: 1.74 usec per loop
$ ./python_work -m timeit -s "import posixpath" "posixpath.relpath('/', start='/home/account')"
200000 loops, best of 5: 1.53 usec per loop
```

Windows - posixpath - 1.123-1.136 speedup

```text
> .\python_main -m timeit -s "import posixpath" "posixpath.relpath('foo', start='/home/account')"
50000 loops, best of 5: 4.39 usec per loop
> .\python_work -m timeit -s "import posixpath" "posixpath.relpath('foo', start='/home/account')"
100000 loops, best of 5: 3.91 usec per loop

> .\python_main -m timeit -s "import posixpath" "posixpath.relpath('/', start='/home/account')"
100000 loops, best of 5: 3.1 usec per loop
> .\python_work -m timeit -s "import posixpath" "posixpath.relpath('/', start='/home/account')"
100000 loops, best of 5: 2.73 usec per loop
```

Linux - ntpath - 1.006-1.007 slowdown (no change)

```text
$ ./python_main -m timeit -s "import ntpath" "ntpath.relpath('C:foo', start='C:/Users/account')"
50000 loops, best of 5: 8.63 usec per loop
$ ./python_work -m timeit -s "import ntpath" "ntpath.relpath('C:foo', start='C:/Users/account')"
50000 loops, best of 5: 8.69 usec per loop

$ ./python_main -m timeit -s "import ntpath" "ntpath.relpath('C:/', start='C:/Users/account')"
50000 loops, best of 5: 6.29 usec per loop
$ ./python_work -m timeit -s "import ntpath" "ntpath.relpath('C:/', start='C:/Users/account')"
50000 loops, best of 5: 6.33 usec per loop
```

**Comment by nineteendo** [comment]
Could you also check with `start` not set?

**Comment by eryksun** [comment]
> Could you also check with `start` not set or a relative path?

Sure, except testing `ntpath.relpath("C:/")` isn't meaningful on POSIX. It fails because the "C:" drive is a different device namespace.

Windows - ntpath - 1.201-1.269 speedup

```text
> .\python_main -m timeit -s "import os,ntpath; os.chdir('C:/Users')" "ntpath.relpath('foo')"
50000 loops, best of 5: 4.84 usec per loop
> .\python_work -m timeit -s "import os,ntpath; os.chdir('C:/Users')" "ntpath.relpath('foo')"
50000 loops, best of 5: 4.03 usec per loop

> .\python_main -m timeit -s "import os,ntpath; os.chdir('C:/Users')" "ntpath.relpath('C:/')"
50000 loops, best of 5: 4.11 usec per loop
> .\python_work -m timeit -s "import os,ntpath; os.chdir('C:/Users')" "ntpath.relpath('C:/')"
100000 loops, best of 5: 3.24 usec per loop
```

Linux - posixpath - 1.309-1.434 speedup

```text
$ ./python_main -m timeit -s "import os,posixpath; os.chdir('/home')" "posixpath.relpath('foo')"
100000 loops, best of 5: 2.94 usec per loop
$ ./python_work -m timeit -s "import os,posixpath; os.chdir('/home')" "posixpath.relpath('foo')"
100000 loops, best of 5: 2.05 usec per loop

$ ./python_main -m timeit -s "import os,posixpath; os.chdir('/home')" "posixpath.relpath('/')"
100000 loops, best of 5: 2.12 usec per loop
$ ./python_work -m timeit -s "import os,posixpath; os.chdir('/home')" "posixpath.relpath('/')"
200000 loops, best of 5: 1.62 usec per loop
```

Windows - posixpath - 1.060 slowdown, 1.019 speedup

```text
> .\python_main -m timeit -s "import os,posixpath; os.chdir('C:/Users')" "posixpath.relpath('foo')"
50000 loops, best of 5: 4.01 usec per loop
> .\python_work -m timeit -s "import os,posixpath; os.chdir('C:/Users')" "posixpath.relpath('foo')"
50000 loops, best of 5: 4.25 usec per loop

> .\python_main -m timeit -s "import os,posixpath; os.chdir('C:/Users')" "posixpath.relpath('/')"
100000 loops, best of 5: 3.22 usec per loop
> .\python_work -m timeit -s "import os,posixpath; os.chdir('C:/Users')" "posixpath.relpath('/')"
100000 loops, best of 5: 3.16 usec per loop
```

Linux - ntpath - 1.004 speedup (no change)

```text
$ ./python_main -m timeit -s "import os,ntpath; os.chdir('/home')" "ntpath.relpath('foo')"
50000 loops, best of 5: 8.23 usec per loop
$ ./python_work -m timeit -s "import os,ntpath; os.chdir('/home')" "ntpath.relpath('foo')"
50000 loops, best of 5: 8.2 usec per loop
```

**Comment by erlend-aasland** [comment]
Thanks for running benchmarks, @eryksun 🙏

**Comment by nineteendo** [comment]
What are your thoughts on the updated code?

**Comment by zooba** [comment]
Thanks!

**Pull request merged**